### PR TITLE
Add `--remove-all-notebook-metadata` flag

### PR DESCRIPTION
**Pull request by yasirroni**
Add new argument for clean and filter "-M" or "--remove-notebook-metadata" to "remove notebook metadata.

```python
    clean_parser.add_argument(
        "-M",
        "--remove-notebook-metadata",
        action="store_true",
        help="remove notebook metadata",
    )
```

### CODE REVIEW

**Comment by srstevenson** [comment]
As this is branched from #161, I'll wait until that's complete before reviewing this.

**Comment by yasirroni** [comment]
https://github.com/srstevenson/nb-clean/pull/161 is merged. Please review this branch. I will rebase this branch.

**Comment by yasirroni** [comment]
After some reading it seems current behavior of nb-clean is cleaning notebook metadata?

File: README.md [hunk 1]
```diff
@@ -262,16 +262,22 @@ that of `nb-clean` >=2.0.0.
...
+| Check notebook (ignore non-empty cells)                      | `nb-clean check -i/--input notebook.ipynb -e/--remove-empty`        | `nb-clean check notebook.ipynb -e/--remove-empty-cells`                  |
+| Check notebook (ignore cell metadata)                        | `nb-clean check -i/--input notebook.ipynb -m/--preserve-metadata`   | `nb-clean check notebook.ipynb -m/--preserve-cell-metadata`              |
+| Check notebook (ignore `tags` and `special` cell metadata)   |                                                                     | `nb-clean check notebook.ipynb -m/--preserve-cell-metadata tags special` |
+| Check notebook (ignore cell outputs)                         |                                                                     | `nb-clean check notebook.ipynb -o/--preserve-cell-outputs`               |
+| Check notebook (ignore notebook metadata)                    |                                                                     | `nb-clean check notebook.ipynb -M/--remove-notebook-metadata`            |
```
Comment (srstevenson) [inline README.md:278]: This doesn't look to be fully rebased: there aren't short flags such as `-i` in the nb-clean 1.6.0 column on `main`.

File: src/nb_clean/__init__.py [hunk 2]
```diff
@@ -82,6 +82,7 @@ def git_attributes_path() -> pathlib.Path:
 def add_git_filter(
     *,
     remove_empty_cells: bool = False,
+    remove_notebook_metadata: bool = False,
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:85]: We now have two arguments to this (and other functions), `remove_notebook_metadata` and `preserve_notebook_metadata`, which both default to `False` which is semantically inconsistent: we need to renconcile this in a way that maintains backward compatibility for existing users of `--preserve-notebook-metadata`.

File: src/nb_clean/__init__.py [hunk 3]
```diff
@@ -227,7 +234,11 @@ def check_notebook(
                 print(f"{prefix}: outputs")
                 is_clean = False
 
-    if not preserve_notebook_metadata:
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:230]: Following up on my comment above about having both `remove_notebook_metadata` and `preserve_notebook_metadata`, by removing this `--preserve-notebook-metadata` is no longer observed which breaks existing uses.

File: src/nb_clean/__init__.py [hunk 4]
```diff
@@ -274,6 +288,12 @@ def clean_notebook(
     if remove_empty_cells:
         notebook.cells = [cell for cell in notebook.cells if cell["source"]]
 
+    if remove_notebook_metadata:
+        notebook.metadata = {}
+    else:
+        with contextlib.suppress(KeyError):
+            del notebook["metadata"]["language_info"]["version"]
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:295]: The `else` branch here is a duplicate of the code on line 317, but that code is run under different conditions: it's run if `preserve_notebook_metadata` is False, whereas the code here is run if `remove_notebook_metadata` is False.

**Review by srstevenson — CHANGES_REQUESTED** [review]
Before we consider merging this, we need to think more about the approach and how to maintain compatibility for existing users.

With this change, there are `--preserve-notebook-metadata` and `--remove-notebook-metadata` arguments which sound as though they are the inverse of one another but do different things and are toggled independently. This is confusing for users, and also for me to maintain.

`--preserve-notebook-metadata`  only checks for and removes the `language_info.version` key from the metadata, whereas the new `--remove-notebook-metadata` argument removes all metadata keys. We need the names to disambiguate this: for example, by renaming `--remove-notebook-metadata` to `--remove-all-notebook-metadata` or similar.

Even if it were named differently, I'm not keen to maintain two flags so easily confused with each other. Perhaps we should deprecate `--preserve-notebook-metadata`? This can be achieved in a sequence of steps across future releases:

- from this change, using `--preserve-notebook-metadata` prints a deprecation warning that the feature will be removed in future
- after the deprecation warning has been present for some time, the flag becomes a no-op that prints a warning saying the feature was removed
- eventually, we remove the flag entirely and passing it will result in an argument parsing error

In addition, please can you run the linters and test suite with `poetry run poe check` and address the issues raised?

**Comment by yasirroni** [comment]
All updated. Please review.

File: README.md [hunk 5]
```diff
@@ -192,6 +192,13 @@ To preserve notebook metadata, such as language version, use:
 nb-clean add-filter --preserve-notebook-metadata
 ```
 
+By default, `nb-clean` will not delete all notebook metadata. To completely
```
Comment (srstevenson) [inline README.md:195]: GitHub doesn't allow commenting on lines not included within the diff in a PR, but we also need to mention `--remove-all-notebook-metadata` where we list the available options in the "Checking" and "Cleaning (interactive)" sections above, starting at lines 61 and 116. In those, we should also make clear that `--preserve-notebook-metadata` only checks for the presence of the `language_info.version` key, to explain how it differs from `--remove-all-notebook-metadata`.

File: README.md [hunk 6]
```diff
@@ -262,16 +269,17 @@ that of `nb-clean` >=2.0.0.
...
+| ------------------------------------------------------------ | ---------------------------------------------------------------- | -----------------------------------------------------------          |
+| Clean notebook                                               | `nb-clean clean --input notebook.ipynb \| sponge notebook.ipynb` | `nb-clean clean notebook.ipynb`                                      |
+| Clean notebook (remove empty cells)                          | `nb-clean clean --input notebook.ipynb --remove-empty`           | `nb-clean clean --remove-empty-cells notebook.ipynb`                 |
+| Clean notebook (preserve cell metadata)                      | `nb-clean clean -i/--input notebook.ipynb -m/--preserve-metadata`   | `nb-clean clean notebook.ipynb -m/--preserve-cell-metadata`       |
+| Clean notebook (remove all notebook metadata)                |                                                                     | `nb-clean clean notebook.ipynb -M/--remove-all-notebook-metadata` |
```
Comment (srstevenson) [inline README.md:277]: We don't need to include this in the migration table, as the feature wasn't present in nb-clean 1.6.0 so there's no existing command to migrate from.

File: src/nb_clean/__init__.py [hunk 7]
```diff
@@ -93,6 +94,8 @@ def add_git_filter(
     ----------
     remove_empty_cells : bool, default False
         If True, remove empty cells.
+    remove_all_notebook_metadata : bool, default False
+        If True, remove all notebooks metadata
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:98]:
```suggestion
        If True, remove all notebook metadata.
```

File: src/nb_clean/__init__.py [hunk 8]
```diff
@@ -173,6 +180,8 @@ def check_notebook(
         The notebook.
     remove_empty_cells : bool, default False
         If True, also check for the presence of empty cells.
+    remove_all_notebook_metadata : bool, default False
+        If True, also check for the presence of nonempty notebook metadata.
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:184]:
```suggestion
        If True, also check for the presence of any notebook metadata.
```

File: src/nb_clean/__init__.py [hunk 9]
```diff
@@ -128,6 +131,9 @@ def add_git_filter(
     if preserve_notebook_metadata:
         command.append("--preserve-notebook-metadata")
 
+    if remove_all_notebook_metadata:
+        command.append("--remove-all-notebook-metadata")
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:135]: As flagged by Codecov, this addition isn't covered by tests.

File: src/nb_clean/__init__.py [hunk 10]
```diff
@@ -159,6 +165,7 @@ def check_notebook(
     notebook: nbformat.NotebookNode,
     *,
     remove_empty_cells: bool = False,
+    remove_all_notebook_metadata: bool = False,
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:168]: At the start of the function body, here and for all other functions that take both arguments, let's check that `remove_all_notebook_metadata` and `preserve_notebook_metadata` aren't both `True`, and raise an appropriate exception if so (this combination is impossible to satisfy).

File: src/nb_clean/__init__.py [hunk 11]
```diff
@@ -253,6 +267,8 @@ def clean_notebook(
         The notebook.
     remove_empty_cells : bool, default False
         If True, remove empty cells.
+    remove_all_notebook_metadata : bool, default False
+        If True, remove notebook metadata.
```
Comment (srstevenson) [inline src/nb_clean/__init__.py:271]:
```suggestion
        If True, remove all notebook metadata.
```

File: src/nb_clean/cli.py [hunk 12]
```diff
@@ -181,6 +184,12 @@ def parse_args(args: list[str]) -> argparse.Namespace:
     add_filter_parser.add_argument(
         "-e", "--remove-empty-cells", action="store_true", help="remove empty cells"
     )
+    add_filter_parser.add_argument(
+        "-M",
+        "--remove-all-notebook-metadata",
+        action="store_true",
+        help="remove notebook metadata",
```
Comment (srstevenson) [inline src/nb_clean/cli.py:191]:
```suggestion
        help="remove all notebook metadata",
```

File: src/nb_clean/cli.py [hunk 13]
```diff
@@ -226,6 +235,12 @@ def parse_args(args: list[str]) -> argparse.Namespace:
     check_parser.add_argument(
         "-e", "--remove-empty-cells", action="store_true", help="check for empty cells"
     )
+    check_parser.add_argument(
+        "-M",
+        "--remove-all-notebook-metadata",
+        action="store_true",
+        help="check for notebook metadata",
```
Comment (srstevenson) [inline src/nb_clean/cli.py:242]:
```suggestion
        help="check for any notebook metadata",
```

File: src/nb_clean/cli.py [hunk 14]
```diff
@@ -262,6 +277,12 @@ def parse_args(args: list[str]) -> argparse.Namespace:
     clean_parser.add_argument(
         "-e", "--remove-empty-cells", action="store_true", help="remove empty cells"
     )
+    clean_parser.add_argument(
+        "-M",
+        "--remove-all-notebook-metadata",
+        action="store_true",
+        help="remove notebook metadata",
```
Comment (srstevenson) [inline src/nb_clean/cli.py:284]:
```suggestion
        help="remove all notebook metadata",
```

File: tests/test_check_notebook.py [hunk 15]
```diff
@@ -175,3 +175,36 @@ def test_check_notebook_preserve_execution_counts(
...
+    ],
+)
+def test_check_notebook_remove_all_notebook_metadata(
+    notebook_name: str,
+    *,
+    remove_all_notebook_metadata: bool,
+    is_clean: bool,
+    request: pytest.FixtureRequest,
+) -> None:
+    """Test nb_clean.clean_notebook when removing all notebooks metadata.
```
Comment (srstevenson) [inline tests/test_check_notebook.py:198]:
```suggestion
    """Test nb_clean.clean_notebook when removing all notebook metadata.
```

File: tests/test_check_notebook.py [hunk 16]
```diff
@@ -175,3 +175,36 @@ def test_check_notebook_preserve_execution_counts(
...
+
+    The test with `("clean_notebook_with_notebook_metadata", False, True)` is False due
+    to `clean_notebook_with_notebook_metadata` is not clean. That is because it contains
+    `language_info.version` captured by --preserve-notebook-metadata.
```
Comment (srstevenson) [inline tests/test_check_notebook.py:202]: This is a useful comment! Suggested tweak to wording:
```suggestion
    The test with `("clean_notebook_with_notebook_metadata", False, True)` is False due
    to `clean_notebook_with_notebook_metadata` containing `language_info.version` detected
    when `preserve_notebook_metadata=False`.
```

File: tests/test_clean_notebook.py [hunk 17]
```diff
@@ -119,3 +119,14 @@ def test_clean_notebook_preserve_execution_counts(
         nb_clean.clean_notebook(dirty_notebook, preserve_execution_counts=True)
         == clean_notebook_with_counts
     )
+
+
+def test_clean_notebook_remove_all_notebook_metadata(
+    dirty_notebook: nbformat.NotebookNode,
+    clean_notebook_without_notebook_metadata: nbformat.NotebookNode,
+) -> None:
+    """Test nb_clean.clean_notebook when preserving cell outputs."""
```
Comment (srstevenson) [inline tests/test_clean_notebook.py:128]:
```suggestion
    """Test nb_clean.clean_notebook when removing all notebook metadata."""
```

**Review by srstevenson — CHANGES_REQUESTED** [review]
Thanks, I've made another round of review -- we're making good progress.

**Comment by yasirroni** [comment]
All resolved with 100% coverage. Please review.

**Review by srstevenson — APPROVED** [review]
LGTM, thanks for perservering with this.

**Comment by srstevenson** [comment]
This change is included in the [nb-clean 3.3.0](https://pypi.org/project/nb-clean/3.3.0/) release.

**Pull request merged**